### DISCUSSION TITLE
After Updating to PrimeFaces 13 PDF Viewer Displays Same Report Repeatedly

### DISCUSSION THREAD

**Discussion by NCIDevRonR** [primefaces/community#52] [PrimeFaces] [answered]
I have a report generation and viewing sequence which has worked flawlessly in PrimeFaces 6 for years in production.  Upon updating to PrimeFaces 13, the last part of the sequence is now broken, and I'm not sure what to change to get it working again.

The user clicks a button to generate the report after selecting the criteria.  The Jasper Reports code generates the report, and I can verify this by viewing the named and timestamped report files produced on the server.

After the report is generated, a new streaming PDF object is created for the report just requested, and an onscreen "poll" object downloads the just-created streaming object into the screen's PDF viewer.

The PDF viewer appears, and displays the requested report for the user, then the user clicks a button to close the PDF viewer.

What has changed since we've updated to PrimeFaces 13 is if the user requests a succession of different report types and criteria, although each requested report is generated on the back-end, only the first requested report displays each time, whereas in PrimeFaces 6, each requested report is generated and displayed in turn.

I'm including a sizeable amount of code here to show how the entire sequence is set up to behave, including a call to the new DefaultStreamedContent.builder() which replaced the parameterized constructor.  Looking at this, does anything jump out at you as a possible cause for why the PDF viewer only shows the first generated report repeatedly in PrimeFaces 13?  Can anyone recommend a way to get the PDF viewer to display each generated report in turn, as it does in PrimeFaces 6?
[Issue with Jasper Reports-20230915.txt](https://github.com/primefaces/community/files/12622917/Issue.with.Jasper.Reports-20230915.txt)

**Comment by melloware** [comment]
Can you post your Controller code that does the `DefaultStreamedContent.builder()` so we can see?  What scope is it?

**Reply by NCIDevRonR** [reply]
I've included it in the code snippet.  It's near the bottom of the file.

It's "public static StreamedContent getFileSystemPDF(String filePath, String simpleFilename)" defined in a Common_Util class, and called from inside "private StreamedContent createStream(String filePath)" in my JasperController class.

**Reply by melloware** [reply]
Yep but your code is cut off I can't see if it is `@RequestScoped` or not?

**Reply by NCIDevRonR** [reply]
Ah!  Ok.  My JasperController class is @SessionScoped, and my Common_Util class is not annotated.

**Reply by melloware** [reply]
yeah that won't work I thought I think you want that StreamedContent to be `@RequestScoped` like all the PF FileDownload examples.

**Reply by melloware** [reply]
I think its caching your in memory `private StreamContent` field and serving the same one over and over instead of a RequestScoped would construct it each time and serve that instance.

**Reply by NCIDevRonR** [reply]
I've just looked at a PF file download example, and I (think I) see what you mean.  I'll try something more in line with that next week and let you know if it works.

**Reply by NCIDevRonR** [reply]
Well it certainly _sounded_ straight-forward, however for me today the _fun_ was in the details, some of which I'm sure I've gotten wrong.

I have also broken the document viewer, which was at least working before, to where it displays a miniature of the main screen instead of the document.  A breakpoint in my new JasperDownloader's getFile() method is never reached.

Is what I'm missing here obvious?

Please see the attached code update.  Everything else is as it was.

[Issue with Jasper Reports-20230918.txt](https://github.com/primefaces/community/files/12663426/Issue.with.Jasper.Reports-20230918.txt)

**Reply by melloware** [reply]
You are mixing paradigms you can't have a constructor `jdl = new JasperDownloader(filePath, simpleFilename);` and have it be @ReuestScoped because CDI is going to create a new instance of `JasperDownloader`.    So you are constructing an instance in your SessionScoped bean and hoping its the instance being created by CDI.  

Is there a reason your `JasperController` is `SessionScoped`?  It feels like you could move a lot more of this code into the `RequestScoped` bean?  I can't see all your code but I assume there is some reason you are storing it in Session.  Another thing you can try is this if you really don't want to change this code.

In your JasperController try this.

```java
public StreamedContent getFile() {
        String simpleFilename = filePath.substring(filePath.lastIndexOf(FILE_SEPARATOR) + 1);
            return DefaultStreamedContent.builder()
                    .stream(() -> new FileInputStream(filePath))
                    .contentType("application/pdf")
                    .name(simpleFilename)
                    .build();
    }
```

its ugly but it gets you away from storing a variable?

**Reply by NCIDevRonR** [reply]
I've been experimenting with creating a separate @RequestScoped JasperDownloader, because when I switched my JasperController to @RequestScoped, my report viewer looked like this.
![ReportViewer with JC ReqScope-20230919-001](https://github.com/primefaces/community/assets/78611742/6c34d814-4709-4a2d-8368-caee56a11d63)
...with nothing where the viewable report window should be.  I just assumed this meant the JasperController needed to remain @SessionScoped as it had been under PrimeFaces 6.
Shouldn't I expect your getFile() recommendation, called from the GUI to behave @SessionScoped within JasperController, just like my current code?  I'll give it a try, but I think that's how it will behave.

**Reply by melloware** [reply]
Not sure you will have to try it.  I just didn't think it needed to be SessionScoped because I can't imagine what you are storing in the session for a report?

**Reply by NCIDevRonR** [reply]
I've added a version of your recommended getFile() method to my JasperController, modified to catch the FileNotFoundException, and tested it with, and without the @RequestScoped annotation.  Both ways, the PDF Viewer behaved session scoped and only displayed the first requested report repeatedly.  I'm including the code snippet.
[Issue with Jasper Reports-20230919.txt](https://github.com/primefaces/community/files/12663366/Issue.with.Jasper.Reports-20230919.txt)
Maybe the fix is to switch the JasperController to @RequestScoped and to correct whatever's preventing the PDF viewer from showing the PDF in this scope.  I'll play around with trying to resolve this today.  Any ideas what might need changing?

**Reply by melloware** [reply]
Hmmm I am stumped on this one without a working reproducer to debug.  But there is clearly something going on that its serving the same file over and over.  Let me know what you find out?

**Reply by NCIDevRonR** [reply]
The fact that you're stumped too makes me feel a little better about how stumped this has me.  I think I'm going to spend some time tomorrow trying to create a nice compact "working reproducer" to share here.  Hopefully I know the pieces of the problem well enough to bring them together so as to make the test app behave oddly in the same way our app is doing.

**Reply by NCIDevRonR** [reply]
I have successfully created a "working reproducer" which both works and misbehaves like our app does under PrimeFaces 13.

**Reply by melloware** [reply]
Can you just wrap this up in a Running PrimeFaces Test project: https://github.com/primefaces/primefaces-test

**Reply by NCIDevRonR** [reply]
I'm not super-experienced with creating forks, but I believe I have successfully done so.  My project is [https://github.com/NCIDevRonR/primefaces-test/tree/pdf-viewer-caching](https://github.com/NCIDevRonR/primefaces-test/tree/pdf-viewer-caching), and it includes the needed PDFs in the "resource" folder.

**Reply by melloware** [reply]
It must be private it won't let me access it. You can download as zip and just put the zip here do NOT include the /target folder!

**Reply by NCIDevRonR** [reply]
I think it'd be preferable if we could make this work using the repo.  It's set up as "public", but it looks like I needed to add you to it as a collaborator.  I've reorganized it a bit, added you as a collaborator, and have updated the link in the comment above.  I'll do the zip file thing if you're still unable to access it through my repo.

**Reply by melloware** [reply]
I see it looking now.

**Reply by melloware** [reply]
OK @NCIDevRonR I untangled the sphaghetti to make it work but I had to make a ton of changes.  I cannt believe this code ever worked (although I do believe you) this code needs a major refactoring and overhaul. Attached is my modified zip of your code:
[pf-pdf.zip](https://github.com/primefaces/community/files/12702857/pf-pdf.zip)

Simply edit 1 line and change to where your files are:
```java
    private static final String LISTING_OUTPUT_PATH = "C:\\temp\\primefaces\\pf-pdf\\src\\main\\webapp\\resources\\";
```

I wasn't about to untangle your dynamic pathing so I just hardcoded the path to where the PDF's are.

**Reply by NCIDevRonR** [reply]
Thank you for looking at this with me.  I'm checking most of your changes into my repo for this reproducer as a step in the right direction, but I do have an observation and a question:

You have replaced my full-sized PDF viewer dialog with an in-line mini PDF viewer, which I don't think would work for our users.  They see the PDF viewer dialog after they have specified their report criteria in another dialog they see first.  The dialog, size and format, make it easy to read and to page through each detailed report.  Also several concurrent users can be requesting variations of the same report at once, hence the timestamps in the filenames.

Was it a matter of preference to switch to the main-window mini-viewer, or was there no way to make the full-sized PDF viewer dialog show the last requested report the same way your new inline viewer does?  (I tried changing the "cache" setting of my dialog-based PDF viewer to 'false' to match yours, and it didn't resolve the issue.)

**Reply by melloware** [reply]
I only did that to make my repeated testing easier to show you that the PDF can work work fine.  I wanted to make a "Minimum Viable Reproducer" so I cut out a lot of your moving parts.  Its always best to show as simple as possible the issue.

**Reply by NCIDevRonR** [reply]
Ok, but I still have a problem, albeit a somewhat smaller problem thanks to your fix.  I have incorporated said fix into another commit to my reproducer repo, so that the PDF viewer resides once again inside my dialog.  Whereas your inline PDF viewer updated when the user clicked the "Display" button, my dialog-based PDF viewer seems to appear in a pre-form-update state.  If I click "refresh" before selecting another listing to display, the correct listing always displays.  Is there a way to accomplish this screen refresh programmatically so the dialog always displays the requested listing?

**Reply by melloware** [reply, marked as answer]
Yes in the Java side you can do `PrimeFaces.current().ajax().update("yourForm")` or whatever you want to update which should probably be your dialog guts?

**Reply by NCIDevRonR** [reply]
Cool!  This resolves my issue (at least in the reproducer).  I have updated the repo with this last fix, and am prepared to consider this case closed.  Thanks for your help.
